plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/UniqueDeclarationChecker.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2010-2024 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver
+
+import org.jetbrains.kotlin.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.FirSession
+import org.jetbrains.kotlin.fir.analysis.checkers.MppCheckerKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.declaration.FirSimpleFunctionChecker
+import org.jetbrains.kotlin.fir.declarations.FirSimpleFunction
+import org.jetbrains.kotlin.fir.declarations.FirVariable
+import org.jetbrains.kotlin.fir.declarations.hasAnnotation
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.expressions.arguments
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+import org.jetbrains.kotlin.fir.symbols.SymbolInternals
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+sealed class UniquenessLevel
+data object Unique : UniquenessLevel()
+data object Shared : UniquenessLevel()
+
+class UniqueDeclarationChecker(private val session: FirSession) : FirSimpleFunctionChecker(MppCheckerKind.Common) {
+
+    private fun getAnnotationId(name: String): ClassId =
+        ClassId(FqName.fromSegments(listOf("org", "jetbrains", "kotlin", "formver", "plugin")), Name.identifier(name))
+
+    private val uniqueId: ClassId = getAnnotationId("Unique")
+
+    private fun getUniqueLevelFromExpression(expression: FirExpression, context: Map<FirVariable, UniquenessLevel>): UniquenessLevel {
+        return Shared
+    }
+
+    @OptIn(SymbolInternals::class)
+    override fun check(declaration: FirSimpleFunction, context: CheckerContext, reporter: DiagnosticReporter) {
+        val errorCollector = ErrorCollector()
+        try {
+            // uniquenessContext is a map from variable to its uniqueness level
+            val uniquenessContext: MutableMap<FirVariable, UniquenessLevel> = mutableMapOf()
```

Comment:
I'd recommend making a `formver.uniqueness` subdirectory and then calling out to something there for context stuff like this.  The checker will grow big quickly otherwise.